### PR TITLE
Add checkbox to delete pull branch after successful merge

### LINKED ISSUE #13568: Automatic Deletion of Unprotected branches on PR merge with toggleable switch
- Gitea version (or commit ref): 1.14.0+dev-208-g6280af032
- Git version: 2.29.0
- Operating system: Linux Ubuntu 16.04
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No (not a bug.)
- Log gist:

## Description

* Toggleable switch at project level for use
* if an **unprotected** branch is merged while switch is set to true, that branch gets deleted
* protected branches are unaffected

### PR DESCRIPTION
**Pull request by jpraet**
Fixes #13568

### CODE REVIEW

**Comment by jpraet** [comment]
Sometimes these errors are logged:

```
2021/06/02 09:12:34 ...es/pull/temp_repo.go:144:createTemporaryRepo() [E] Unable to fetch head_repo head branch [jpraet/test:jpraet-patch-2 -> tracking in C:\gitea\data\tmp\local-repo\pull.git592031210]: exit status 128:

        fatal: Couldn't find remote ref refs/heads/jpraet-patch-2

2021/06/02 09:12:34 ...vices/pull/update.go:91:GetDiverging() [E] CreateTemporaryPath: Unable to fetch head_repo head branch [jpraet/test:jpraet-patch-2 -> tracking in tmpBasePath]: exit status 128

        fatal: Couldn't find remote ref refs/heads/jpraet-patch-2

2021/06/02 09:12:34 ...ervices/pull/pull.go:305:func1() [E] GetDiverging: Unable to fetch head_repo head branch [jpraet/test:jpraet-patch-2 -> tracking in tmpBasePath]: exit status 128

        fatal: Couldn't find remote ref refs/heads/jpraet-patch-2

```

When a PR gets merged, a TestPullRequest task is scheduled asynchrounously:

https://github.com/go-gitea/gitea/blob/894742278154781ba4cb347ac4c7efd7d6dcd1c8/services/pull/merge.go#L55

If the branch gets deleted before this task has completed, you will get these errors in the logs.

What is the purpose of this task?

**Comment by jpraet** [comment]
It looks like it's not the task scheduled from merge.go that is causing the problem, because if the PR was successfully merged that one doesn't do anything.

The problematic one is scheduled here:

https://github.com/go-gitea/gitea/blob/6a7bf974bef6257517e39d9979b4ebac8c35176a/services/repository/push.go#L141

as a result of the push to the target branch, which happens before the PR is marked as merged.

**Comment by lunny** [comment]
It seems CI is PASS

**Comment by jpraet** [comment]
Yes it works fine functionally, but the errors in the logs are unpleasant. I made a change to detect this scenario and log it as a warning instead of error, not sure if that's the right approach?

```
2021/06/03 22:38:35 ...ervices/pull/pull.go:306:func1() [W] Cannot test PR my-repo/45: head_branch jpraet-patch-1 no longer exists
```

**Comment by lunny** [comment]
When delete a branch, gitea should close all related PRs, then the check will not continue and the error log will not occur.

So I think this is a bug of `cleanupbranch` we haven't found before. It should close all related opened PRs.

**Comment by jpraet** [comment]
> 
> 
> When delete a branch, gitea should close all related PRs, then the check will not continue and the error log will not occur.
> 
> So I think this is a bug of `DeleteBranch` we haven't found before. It should close all related opened PRs.
> 
> UPDATE:
> 
> I sent #16067 to fix the bug.

I don't think that's the problem here. The PR is already marked as merged and closed before I call `CleanupPullRequest()`. And as I commented in your PR: it looks like `CloseBranchPulls()` is already being called when deleting a branch.

As far as I can tell the issue is the asynchronous `AddTestPullRequestTask` goroutine that is automatically scheduled when the merge commit is pushed to the target branch. While that task is running, we are deleting the branch.

**Comment by jpraet** [comment]
Should this also be supported through the API? To me, it seems like this compound operation of merging a PR and also deleting the source branch mainly makes sense as a UI convenience.

**Comment by lunny** [comment]
It's OK from me to allow API to do that.

File: routers/web/repo/pull.go [hunk 1]
```diff
@@ -965,6 +965,11 @@ func MergePullRequest(ctx *context.Context) {
 	}
 
 	log.Trace("Pull request merged: %d", pr.ID)
+
+	if form.DeleteBranchAfterMerge {
+		cleanUpPullRequest(ctx, false)
```
Comment (zeripath) [inline routers/web/repo/pull.go]: You already have the pull request, and git repo open which means that at least until 1204 is already done.

Comment (jpraet) [reply routers/web/repo/pull.go]: I'll see if I can do some cleanup to avoid these unnecessary operations.
How can I access the already opened repo of the PR head?

File: routers/api/v1/repo/pull.go [hunk 2]
```diff
@@ -876,6 +878,27 @@ func MergePullRequest(ctx *context.APIContext) {
 	}
 
 	log.Trace("Pull request merged: %d", pr.ID)
+
+	if form.DeleteBranchAfterMerge {
+		if err := repo_service.DeleteBranch(ctx.User, ctx.Repo.Repository, ctx.Repo.GitRepo, pr.HeadBranch); err != nil {
```
Comment (jpraet) [inline routers/api/v1/repo/pull.go]: After reflecting on https://github.com/go-gitea/gitea/pull/16049#discussion_r660996140, I think this is wrong.
This only works if the pr head branch is in the same repo.
We need to delete the branch on pr.HeadRepo instead of ctx.Repo.Repository.

File: services/pull/pull.go [hunk 3]
```diff
@@ -303,7 +303,11 @@ func AddTestPullRequestTask(doer *models.User, repoID int64, branch string, isSy
 		for _, pr := range prs {
 			divergence, err := GetDiverging(pr)
 			if err != nil {
-				log.Error("GetDiverging: %v", err)
+				if models.IsErrBranchDoesNotExist(err) && !git.IsBranchExist(pr.HeadRepo.RepoPath(), pr.HeadBranch) {
+					log.Warn("Cannot test PR %s/%d: head_branch %s no longer exists", pr.BaseRepo.Name, pr.IssueID, pr.HeadBranch)
```
Comment (jpraet) [inline services/pull/pull.go:307]: I no longer observe this warning when testing this feature on my local instance, whereas previously I saw it most of the time.
Maybe it has been fixed by another PR. Should I revert these changes to the error handling?

Comment (zeripath) [reply services/pull/pull.go:307]: leave it in.

**Review by zeripath — APPROVED** [review]

**Review by 6543 — APPROVED** [review]
